pkg/queues/amqp_provider.go

```diff
@@ -0,0 +1,189 @@
+	}
+	defer func() {
+		if err != nil {
+			if channel != nil {
+				channel.Close()
+			}
+			if connection != nil {
+				connection.Close()
+			}
+		}
+	}()
+	connection, err = amqp.Dial(r.url)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create connection: %w", err)
+	}
+	channel, err = connection.Channel()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create channel: %w", err)
+	}
+	_, err = channel.QueueDeclare(queueName,
+		true,  // durable
+		false, // auto delete
+		false, // exclusive
+		false, // no wait
+		nil) // args
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to declare queue %s: %w", queueName, err)
+	}
+	ret := &amqpQueue{
+		name:       queueName,
+		connection: connection,
+		channel:    channel,
+		log:        r.log,
+		ctx:        r.ctx,
+		wg:         r.wg,
+	}
+	r.queues = append(r.queues, ret)
+	return ret, nil
+}
+
+func (a *amqpProvider) NewConsumer(queueName string) (Consumer, error) {
+	q, err := a.newQueue(queueName)
+	return q, err
+}
+
+func (a *amqpProvider) NewPublisher(queueName string) (Publisher, error) {
+	q, err := a.newQueue(queueName)
+	return q, err
+}
+
+func (a *amqpProvider) Stop() {
+	a.cancel()
+	for _, q := range a.queues {
+		q.close()
+	}
+}
+
+func (a *amqpProvider) Wait() {
+	a.wg.Wait()
+}
+
+type amqpQueue struct {
+	ctx        context.Context
+	connection *amqp.Connection
+	channel    *amqp.Channel
+	name       string
+	wg         *sync.WaitGroup
+	log        logrus.FieldLogger
+	closed     bool
+}
+
+func (r *amqpQueue) Publish(payload []byte) error {
+	if r.closed {
+		return errors.New("queue is closed")
+	}
+	return r.channel.Publish("",
+		r.name, // queue name
+		false,  // mandatory
+		false,  // immediate
+		amqp.Publishing{
+			DeliveryMode: amqp.Persistent,
+			ContentType:  "text/plain",
+			Body:         payload,
+		})
+}
+
+func (r *amqpQueue) Consume(handler ConsumeHandler) error {
+	msgs, err := r.channel.ConsumeWithContext(r.ctx,
+		r.name,
+		"",    // consumer (exchange)
+		false, // auto ack
+		false, // exclusive
+		false, // no local
+		false, // no wait
+		nil) // args
+	if err != nil {
+		return err
+	}
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		var err error
+		for {
+			select {
+			case <-r.ctx.Done():
+				return
+			case d, ok := <-msgs:
+				requestID := reqid.NextRequestID()
+				reqCtx := context.WithValue(r.ctx, middleware.RequestIDKey, requestID)
+				log := log.WithReqIDFromCtx(reqCtx, r.log)
+				if !ok {
+					if !r.closed {
+						log.Fatal("channel was closed by AMQP provider")
+					}
+					return
+				}
+				if err = handler(reqCtx, d.Body, log); err != nil {
+					log.WithError(err).Errorf("failed to consume message: %s", string(d.Body))
+				}
+				if err = d.Ack(false); err != nil {
+					log.WithError(err).Errorf("failed to acknowledge message")
+				}
+			}
+
+		}
+	}()
+	return nil
+}
+
+func (a *amqpQueue) close() {
+	a.closed = true
+	if a.channel != nil {
+		a.channel.Close()
```

Comment:
question: what happens if close is called more than once is it idempotent?

Reply:
Good point.  I believe it should be OK but I will check.

Reply:
Made sure it is called only once